### PR TITLE
lib: use Node.js net lib and reject malformed addresses

### PR DESCRIPTION
**Pull request by ouuan**
See #143. Given the current API of this library, I believe that it's not suitable to support malformed addresses like `0x7f.1`. It's hard to parse them correctly in all cases. It's better to throw errors for them.

This PR uses the Node.js `net` library so it should be reliable.

I suggest releasing v3 after merging this, because:

1. `net.SocketAddress` is added in Node v15.14.0 and v14.18.0. I think it's reasonable to drop support for older Node versions because Node 12 has reached end-of-life since 2022-04-30. If the user cares about security, they should use a new Node version.
2. I would consider throwing new errors as a breaking change.
3. Dropping support for malformed addresses could also be a breaking change.

We could try to (partially) fix the security concerns in a compatible way in v2 (and v1) in another pull request.

### CODE REVIEW

**Comment by abhishek-parative** [comment]
Is this getting merged in?

**Comment by ouuan** [comment]
The maintainer is very inactive. He has no obligation to continue maintaining this library, but I'd suggest switching to a different actively maintained library.

**Comment by abhishek-parative** [comment]
@indutny Can you please review/merge this PR?

**Review by LLotme — APPROVED** [review]
gh pr checkout 144

File: lib/ip.js [hunk 1]
```diff
@@ -82,15 +83,28 @@ ip.toString = function (buff, offset, length) {
...
-  return ipv4Regex.test(ip);
+ip.isV6Format = net.isIPv6;
+
+ip.isValid = function (addr) {
+  return net.isIP(addr) !== 0;
 };
 
-ip.isV6Format = function (ip) {
-  return ipv6Regex.test(ip);
+ip.normalizeStrict = function (addr) {
```
Comment (ljharb) [inline lib/ip.js:94]: (this makes the PR semver-minor)

File: lib/ip.js [hunk 2]
```diff
@@ -82,15 +83,28 @@ ip.toString = function (buff, offset, length) {
...
+  case 4:
+    family = 'ipv4';
+    break;
+  case 6:
+    family = 'ipv6';
+    break;
+  default:
+    throw new Error(`Invalid ip address: ${addr}`);
+  }
+  const { address } = new net.SocketAddress({ address: addr, family });
```
Comment (ljharb) [inline lib/ip.js:106]: this, however, would make it semver-major, both for the destructuring syntax (`ip` has no engines field, so it has to support every node version), and because net.SocketAddress wasn't added until node v14.18.0.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -5,12 +5,12 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        node: [ 12, 14, 16, 18 ]
+        node: [ 14, 16, 18, 20 ]
```
Comment (ljharb) [inline .github/workflows/test.yml:8]: 12 should be kept, and since there's no engines field, older nodes should be added too.

**Review by ljharb — CHANGES_REQUESTED** [review]

File: lib/ip.js (hunk 2 above)
Comment (ouuan) [reply lib/ip.js:106]: I mentioned this in the PR body.

Comment (ljharb) [reply lib/ip.js:106]: ah, so you did :-)

if there's a nonbreaking way to solve the underlying issue, it seems better to me to PR that in first (as you mentioned in the OP), since that way the vuln will be fixed for the greatest number of users.

**Review by LLotme — COMMENTED** [review]
gh pr checkout 144

**Comment by sfc-gh-ksaket** [comment]
Is this PR getting merged?

**Comment by rdebeasi** [comment]
I came across a couple of maintained forks of `node-ip`:

- https://github.com/eggjs/node-ip
- https://github.com/zaptic/neoip

Both forks are a few months old, so there's not enough activity for me to tell whether they're likely to be maintained in the long term. That said, they may be a good solution for folks who aren't yet ready to switch to an alternative.

**Comment by charIeszhao** [comment]
Hi, when can we expect a new release that includes this patch? @ljharb

**Comment by ljharb** [comment]
@charIeszhao why are you pinging me? i'm not a maintainer here.

**Comment by charIeszhao** [comment]
> @charIeszhao why are you pinging me? i'm not a maintainer here.

Sorry, I think that was a mistake. I saw you were the reviewer so I assumed you were the maintainer. My appologies for the ping.
